File: 0690-employee-importance/0690-employee-importance.java
/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        int result = 0;

        HashMap<Integer, Employee> map = new HashMap<>();

        for(Employee emp: employees){
            map.put(emp.id, emp);
        }    

        Queue<Integer> q = new LinkedList<>();

        q.add(id);

        while(!q.isEmpty()){
            Employee currEmp = map.get(q.poll());
            result += currEmp.importance;

            for(int subemp: currEmp.subordinates){
                q.add(subemp);
            }
        }

        return result;
    }

}

// BFS
// Time Complexity: O(N)
// Space Complexity: O(N)

File: 0690-employee-importance/0690-employee-importance.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        result = 0

        hashMap = dict()

        for emp in employees:
            hashMap[emp.id] = emp
        
        queue = deque()

        queue.append(id)

        while queue:
            currEmp = hashMap.get(queue.popleft())

            result += currEmp.importance

            for emp in currEmp.subordinates:
                queue.append(emp)
        
        return result


# //BFS
# // Time Complexity: O(N)
# // Space Complexity: O(N)
